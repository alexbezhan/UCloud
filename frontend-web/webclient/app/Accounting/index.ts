import Usage from "Accounting/Usage";

export {Usage};