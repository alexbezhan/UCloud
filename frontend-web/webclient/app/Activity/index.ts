import { ActivityReduxObject } from "DefaultObjects";
import { ScrollRequest } from "Scroll";

export enum ActivityType {
    DOWNLOAD = "download",
    UPDATED = "updated",
    DELETED = "deleted",
    FAVORITE = "favorite",
    INSPECTED = "inspected",
    MOVED = "moved"
}

export interface ActivityGroup {
    type: ActivityType;
    newestTimestamp: number;
    numberOfHiddenResults: number | null;
    items: Activity[];
}

export interface Activity {
    type: ActivityType;
    timestamp: number;
    fileId: string;
    username: string;
    originalFilePath: string;
}

export interface ActivityFilter {
    collapseAt?: number;
    type?: ActivityType;
    minTimestamp?: Date;
    maxTimestamp?: Date;
}

export interface FavoriteActivity extends Activity {
    favorite: boolean;
}

export interface MovedActivity extends Activity {
    newName: string;
}


export interface ActivityDispatchProps {
    onMount: () => void;
    fetchActivity: (scroll: ScrollRequest<number>, filter?: ActivityFilter) => void;
    resetActivity: () => void;
    setRefresh: (refresh?: () => void) => void;
    updateFilter: (filter: Partial<ActivityFilter>) => void;
}

export interface ActivityOwnProps {
}

export type ActivityProps = ActivityReduxObject & ActivityDispatchProps & ActivityOwnProps;
