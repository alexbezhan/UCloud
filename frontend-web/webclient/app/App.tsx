import {Client} from "Authentication/HttpClientInstance";
import Core from "Core";
import Header from "Navigation/Header";
import {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import * as React from "react";
import * as ReactDOM from "react-dom";
import {Provider} from "react-redux";
import {BrowserRouter} from "react-router-dom";
import {createGlobalStyle, ThemeProvider} from "styled-components";
import {theme, UIGlobalStyle} from "ui-components";
import {invertedColors} from "ui-components/theme";
import {findAvatar} from "UserSettings/Redux/AvataaarActions";
import {store} from "Utilities/ReduxUtilities";
import {isLightThemeStored, setSiteTheme, toggleCssColors} from "UtilityFunctions";

export function dispatchUserAction(type: typeof USER_LOGIN | typeof USER_LOGOUT | typeof CONTEXT_SWITCH): void {
    store.dispatch({type});
}

export async function onLogin(): Promise<void> {
    const action = await findAvatar();
    if (action !== null) store.dispatch(action);
}

const GlobalStyle = createGlobalStyle`
  ${() => UIGlobalStyle}
`;

Client.initializeStore(store);

function App({children}: React.PropsWithChildren<{}>): JSX.Element {
    const [isLightTheme, setTheme] = React.useState(() => {
        const isLight = isLightThemeStored();
        if (!isLight) toggleCssColors(isLight);
        return isLight;
    });
    const setAndStoreTheme = (isLight: boolean): void => (setSiteTheme(isLight), setTheme(isLight));

    function toggle(): void {
        toggleCssColors(isLightTheme);
        setAndStoreTheme(!isLightTheme);
    }

    return (
        <ThemeProvider theme={isLightTheme ? theme : {...theme, colors: invertedColors}}>
            <>
                <GlobalStyle />
                <BrowserRouter basename="app">
                    <Header toggleTheme={toggle} />
                    {children}
                </BrowserRouter>
            </>
        </ThemeProvider>
    );
}

ReactDOM.render(
    (
        <Provider store={store}>
            <App>
                <Core />
            </App>
        </Provider>
    ),
    document.getElementById("app")
);
