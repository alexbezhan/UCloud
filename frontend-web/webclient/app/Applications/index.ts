import {AnalysisReduxObject, ResponsiveReduxObject} from "DefaultObjects";
import {SortOrder} from "Files";
import {History} from "history";
import {SetStatusLoading} from "Navigation/Redux/StatusActions";
import PromiseKeeper from "PromiseKeeper";
import * as React from "react";
import {SharedFileSystemMount} from "Applications/FileSystems";
import {match} from "react-router";
import {ParameterValues} from "Utilities/ApplicationUtilities";

export interface Analysis {
    name: string;
    checked?: boolean;
    status: string;
    state: AppState;
    jobId: string;
    appName: string;
    appVersion: string;
    createdAt: number;
    modifiedAt: number;
    expiresAt?: number;
    owner: string;
    metadata: ApplicationMetadata
}

export type AnalysesStateProps = AnalysisReduxObject & {responsive: ResponsiveReduxObject}
export type AnalysesProps = AnalysesStateProps & AnalysesOperations;

type FetchJobsOperation = (
    itemsPerPage: number,
    pageNumber: number,
    sortOrder: SortOrder,
    sortBy: RunsSortBy,
    minTimestamp?: number,
    maxTimestamp?: number,
    filter?: AppState
) => void;

export interface AnalysesOperations {
    setLoading: (loading: boolean) => void;
    fetchJobs: FetchJobsOperation;
    onInit: () => void;
    setRefresh: (refresh?: () => void) => void;
    checkAnalysis: (jobId: string, checked: boolean) => void;
    checkAllAnalyses: (checked: boolean) => void;
}

export interface DetailedResultOperations {
    setPageTitle: (jobId: string) => void;
    setLoading: (loading: boolean) => void;
    setRefresh: (refresh?: () => void) => void;
}

export interface DetailedResultProps extends DetailedResultOperations {
    match: match<{jobId: string}>;
    history: History;
}

export interface Application {
    favorite: boolean;
    owner: string;
    createdAt: number;
    modifiedAt: number;
    description: ApplicationDescription;
    tool: ApplicationTool;
    imageUrl: string;
}

interface ApplicationTool {
    owner: string;
    createdAt: number;
    modifiedAt: number;
    description: {
        info: ApplicationInfo
        container: string
        defaultNumberOfNodes: number
        defaultTasksPerNode: number
        defaultAllocationTime: MaxTime
        requiredModules: string[]
        authors: string[]
        title: string
        description: string
        backend: string
        license: string
    };
}

interface ApplicationInfo {
    name: string;
    version: string;
}

export interface ApplicationDescription {
    info: ApplicationInfo;
    tool: ApplicationInfo;
    authors: string[];
    title: string;
    description: string;
    invocation: any[]; // FIXME: Add type
    parameters: ApplicationParameter[];
    outputFileGlobs: string[];
    website?: string;
    resources: {multiNodeSupport: boolean};
    tags: string[];
}

export enum AppState {
    VALIDATED = "VALIDATED",
    PREPARED = "PREPARED",
    SCHEDULED = "SCHEDULED",
    RUNNING = "RUNNING",
    TRANSFER_SUCCESS = "TRANSFER_SUCCESS",
    SUCCESS = "SUCCESS",
    FAILURE = "FAILURE",
    CANCELLING = "CANCELLING"
}

export interface DetailedResultState {
    name: string;
    complete: boolean;
    appState: AppState;
    failedState?: AppState;
    status: string;
    app?: ApplicationMetadata;
    stdout: string;
    stderr: string;
    stdoutLine: number;
    stderrLine: number;
    reloadIntervalId: number;
    promises: PromiseKeeper;
    outputFolder?: string;
    appType?: ApplicationType;
    webLink?: string;
    timeLeft: number | null;
}

export type StdElement = {scrollTop: number, scrollHeight: number} | null

export interface MaxTime {
    hours: number;
    minutes: number;
    seconds: number;
}

export interface MaxTimeForInput {
    hours: number;
    minutes: number;
    seconds: number;
}

export interface JobSchedulingOptionsForInput {
    maxTime: MaxTimeForInput;
    numberOfNodes: number;
    tasksPerNode: number;
    name: React.RefObject<HTMLInputElement>;
}

export interface AdditionalMountedFolder {
    readOnly: boolean;
    ref: React.RefObject<HTMLInputElement>;
    defaultValue?: string;
}

export interface AdditionalPeer {
    nameRef: React.RefObject<HTMLInputElement>;
    jobIdRef: React.RefObject<HTMLInputElement>;
}

export interface RunAppState {
    promises: PromiseKeeper;
    jobSubmitted: boolean;
    initialSubmit: boolean;
    application?: FullAppInfo;
    parameterValues: ParameterValues;
    schedulingOptions: JobSchedulingOptionsForInput;
    favorite: boolean;
    favoriteLoading: boolean;
    mountedFolders: AdditionalMountedFolder[];
    additionalPeers: AdditionalPeer[];
    fsShown: boolean;
    sharedFileSystems: { mounts: SharedFileSystemMount[] };
}

export interface RunOperations extends SetStatusLoading {
    updatePageTitle: () => void;
}

export interface RunAppProps extends RunOperations {
    match: match<{appName: string, appVersion: string}>;
    history: History;
    updatePageTitle: () => void;
}

export interface NumberParameter extends BaseParameter {
    defaultValue: {value: number, type: "double" | "int"} | null;
    min: number | null;
    max: number | null;
    step: number | null;
    type: ParameterTypes.Integer | ParameterTypes.FloatingPoint;
}

export interface BooleanParameter extends BaseParameter {
    defaultValue: {value: boolean, type: "bool"} | null;
    trueValue?: string | null;
    falseValue?: string | null;
    type: ParameterTypes.Boolean;
}

export interface InputFileParameter extends BaseParameter {
    defaultValue: string | null;
    type: ParameterTypes.InputFile;
}

export interface InputDirectoryParameter extends BaseParameter {
    defaultValue: string | null;
    type: ParameterTypes.InputDirectory;
}

export interface TextParameter extends BaseParameter {
    defaultValue: {value: string, type: "string"} | null;
    type: ParameterTypes.Text;
}

export interface PeerParameter extends BaseParameter {
    suggestedApplication: string | null
    type: ParameterTypes.Peer
}

export interface SharedFileSystemParameter extends BaseParameter {
    fsType: "EPHEMERAL" | "PERSISTENT"
    mountLocation: string
    exportToPeers: boolean
    type: ParameterTypes.SharedFileSystem
}

interface BaseParameter {
    name: string;
    optional: boolean;
    title: string;
    description: string;
    unitName?: string | React.ReactNode | null;
    type: string;
    visible?: boolean;
}

export type ApplicationParameter = 
	InputFileParameter | 
	InputDirectoryParameter | 
	NumberParameter | 
	BooleanParameter |
    TextParameter | 
    PeerParameter | 
    SharedFileSystemParameter;

type Invocation = WordInvocation | VarInvocation;

interface WordInvocation {
    type: "word";
    word: string;
}

interface VarInvocation {
    type: "var";
    variableNames: string[];
    prefixGlobal: string;
    suffixGlobal: string;
    prefixVariable: string;
    suffixVariable: string;
    variableSeparator: string;
}

interface Info {name: string; version: string;}
export interface Description {
    info: Info;
    tool: Info;
    authors: string[];
    title: string;
    description: string;
    invocation: Invocation[];
    parameters: ApplicationParameter[];
    outputFileGlobs: [string, string];
    tags: string[];
}

export enum ParameterTypes {
    InputFile = "input_file",
    InputDirectory = "input_directory",
    Integer = "integer",
    FloatingPoint = "floating_point",
    Text = "text",
    Boolean = "boolean",
    Peer = "peer",
    SharedFileSystem = "shared_file_system"
}

export interface SearchFieldProps {
    onSubmit: () => void;
    icon: string;
    placeholder: string;
    value: string;
    loading: boolean;
    onValueChange: (value: string) => void;
}

export interface DetailedApplicationSearchReduxState {
    hidden: boolean;
    appName: string;
    appVersion: string;
    tags: string;
    error?: string;
    loading: boolean;
}

export interface DetailedApplicationOperations {
    setAppName: (n: string) => void;
    setVersionName: (v: string) => void;
    fetchApplicationsFromName: (q: string, i: number, p: number, c?: Function) => void;
    fetchApplicationsFromTag: (t: string, i: number, p: number, c?: Function) => void;
}



// New interfaces
export interface ApplicationMetadata {
    name: string;
    version: string;
    authors: string[];
    title: string;
    description: string;
    tags: string[];
    website?: string;
}

type ApplicationType = "BATCH" | "VNC" | "WEB"

export interface ApplicationInvocationDescription {
    tool: Tool;
    invocation: Invocation[];
    parameters: ApplicationParameter[];
    outputFileGlobs: string[];
    applicationType: ApplicationType;
    shouldAllowAdditionalMounts: boolean
    shouldAllowAdditionalPeers: boolean
    allowMultiNode: boolean
}

interface Resources {
    multiNodeSupport: boolean;
    coreRequirements: number;
    memoryRequirementsMb: number;
    gpuRequirements: number;
    tempStorageRequirementsGb: number;
    persistentStorageRequirementsGb: number;
}

interface Tool {
    name: string;
    version: string;
    tool: ToolReference;
}

interface ToolReference {
    owner: string;
    createdAt: number;
    modifiedAt: number;
    description: ToolDescription;
}

interface NameAndVersion {
    name: string;
    version: string;
}

interface ToolDescription {
    info: NameAndVersion;
    container: string;
    defaultNumberOfNodes: number;
    defaultTasksPerNode: number;
    defaultTimeAllocation: MaxTime;
    requiredModules: string[];
    authors: string[];
    title: string;
    description: string;
    backend: string;
    license: string;
}

export interface WithAppMetadata {
    metadata: ApplicationMetadata;
}

export interface WithAppInvocation {
    invocation: ApplicationInvocationDescription;
}

export interface WithAppFavorite {
    favorite: boolean;
}

export enum RunsSortBy {
    state = "STATE",
    application = "APPLICATION",
    startedAt = "STARTED_AT",
    lastUpdate = "LAST_UPDATE",
    createdAt = "CREATED_AT",
    name = "NAME"
}
export interface WithAllAppTags {
    tags: string[];
}

export type FullAppInfo = WithAppFavorite & WithAppInvocation & WithAppMetadata & WithAllAppTags;
