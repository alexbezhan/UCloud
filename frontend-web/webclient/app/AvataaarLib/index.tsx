/*

  MIT License

  Copyright (c) 2017 Pablo Stanley, Fang-Pen Lin
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.

*/

import * as PropTypes from 'prop-types'
import * as React from 'react'

import Avatar, {AvatarStyle} from './avatar'
import {OptionContext, allOptions} from './options'

export {default as Avatar, AvatarStyle} from './avatar'
export {Option, OptionContext, allOptions} from './options'

import {default as PieceComponent} from './avatar/piece';

export interface Props {
  avatarStyle: string
  style?: React.CSSProperties
  topType?: string
  accessoriesType?: string
  hairColor?: string
  facialHairType?: string
  facialHairColor?: string
  clotheType?: string
  clotheColor?: string
  graphicType?: string
  eyeType?: string
  eyebrowType?: string
  mouthType?: string
  skinColor?: string
  pieceType?: string
  pieceSize?: string
  viewBox?: string
}

export default class AvatarComponent extends React.Component<Props> {
  static childContextTypes = {
    optionContext: PropTypes.instanceOf(OptionContext)
  };
  private optionContext: OptionContext = new OptionContext(allOptions);

  getChildContext() {
    return {optionContext: this.optionContext}
  }

  componentWillMount() {
    this.updateOptionContext(this.props)
  }

  componentWillReceiveProps(nextProps: Props) {
    this.updateOptionContext(nextProps)
  }

  render() {
    const {avatarStyle, style} = this.props;
    return <Avatar avatarStyle={avatarStyle as AvatarStyle} style={style} />
  }

  private updateOptionContext(props: Props) {
    const data: {[index: string]: string} = {};
    for (const option of allOptions) {
      const value = props[option.key];
      if (!value) {
        continue
      }
      data[option.key] = value
    }
    this.optionContext.setData(data)
  }
}

export class Piece extends React.Component<Props> {
  static childContextTypes = {
    optionContext: PropTypes.instanceOf(OptionContext)
  };
  private optionContext: OptionContext = new OptionContext(allOptions);

  getChildContext() {
    return {optionContext: this.optionContext}
  }

  componentWillMount() {
    this.updateOptionContext(this.props)
  }

  componentWillReceiveProps(nextProps: Props) {
    this.updateOptionContext(nextProps)
  }

  render() {
    const {avatarStyle, style, pieceType, pieceSize, viewBox} = this.props;
    return <PieceComponent avatarStyle={avatarStyle as AvatarStyle} style={style} pieceType={pieceType} pieceSize={pieceSize} viewBox={viewBox} />
  }

  private updateOptionContext(props: Props) {
    const data: {[index: string]: string} = {};
    for (const option of allOptions) {
      const value = props[option.key];
      if (!value) {
        continue
      }
      data[option.key] = value
    }
    this.optionContext.setData(data)
  }
}
