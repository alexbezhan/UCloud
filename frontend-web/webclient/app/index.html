<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <title>UCloud</title>
</head>
<body>
    <noscript>
        JavaScript is required to run this page.
    </noscript>
    <div id="app"></div>
</body>
</html>
