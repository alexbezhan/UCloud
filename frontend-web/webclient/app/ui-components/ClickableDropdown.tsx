import {KeyCode} from "DefaultObjects";
import * as React from "react";
import {Icon} from "ui-components";
import * as Text from "ui-components/Text";
import Box from "./Box";
import {Dropdown, DropdownContent} from "./Dropdown";

interface ClickableDropdownState {open: boolean;}
interface ClickableDropdownProps<T> {
    children?: any;
    keepOpenOnClick?: boolean;
    trigger: React.ReactNode;
    fullWidth?: boolean;
    height?: string | number;
    width?: string | number;
    minWidth?: string;
    left?: string | number;
    top?: string | number;
    bottom?: string | number;
    right?: string | number;
    options?: {text: string; value: T}[];
    chevron?: boolean;
    overflow?: string;
    colorOnHover?: boolean;
    squareTop?: boolean;
    keepOpenOnOutsideClick?: boolean;
    onChange?: (value: T) => void;
    onTriggerClick?: () => void;
}

class ClickableDropdown<T extends string> extends React.Component<ClickableDropdownProps<T>, ClickableDropdownState> {
    private ref = React.createRef<HTMLDivElement>();

    constructor(props: Readonly<ClickableDropdownProps<T>>) {
        super(props);
        this.state = {open: false};
        let neither = true;
        if (props.children) neither = false;
        if (!!props.onChange && !!props.options) neither = false;
        if (neither) throw Error("Clickable dropdown must have either children prop or options and onChange");
        document.addEventListener("mousedown", this.handleClickOutside);
        document.addEventListener("keydown", this.handleEscPress);
    }

    public componentWillUnmount = (): void => {
        document.removeEventListener("mousedown", this.handleClickOutside);
        document.removeEventListener("keydown", this.handleEscPress);
    };

    public render(): JSX.Element {
        const {keepOpenOnClick, onChange, onTriggerClick, ...props} = this.props;
        let children: React.ReactNode[] = [];
        if (props.options !== undefined && onChange) {
            children = props.options.map((opt, i) => (
                <Box
                    cursor="pointer"
                    width="auto"
                    key={i}
                    ml="-17px"
                    pl="15px"
                    mr="-17px"
                    onClick={() => onChange!(opt.value)}
                >
                    {opt.text}
                </Box>
            ));
        } else if (props.children) {
            children = props.children;
        }
        const emptyChildren = (React.Children.map(children, it => it) ?? []).length === 0;
        const width = this.props.fullWidth ? "100%" : this.props.width;
        return (
            <Dropdown data-tag="dropdown" ref={this.ref} fullWidth={this.props.fullWidth}>
                <Text.TextSpan cursor="pointer" onClick={() => {onTriggerClick?.(); this.setState(() => ({open: !this.state.open}))}}>
                    {this.props.trigger}{props.chevron ? <Icon name="chevronDown" size=".7em" ml=".7em" /> : null}
                </Text.TextSpan>
                {emptyChildren || !this.state.open ? null : (
                    <DropdownContent
                        overflow={"visible"}
                        squareTop={this.props.squareTop}
                        cursor="pointer"
                        {...props}
                        width={width}
                        hover={false}
                        visible={this.state.open}
                        onClick={() => !keepOpenOnClick ? this.setState(() => ({open: false})) : null}
                    >
                        {children}
                    </DropdownContent>
                )}
            </Dropdown>
        );
    }

    // https://stackoverflow.com/questions/32553158/detect-click-outside-react-component#42234988
    private handleClickOutside = event => {
        if (this.props.keepOpenOnOutsideClick) return;
        if (this.ref.current && !this.ref.current.contains(event.target) && this.state.open) {
            this.setState(() => ({open: false}));
        }
    };

    private handleEscPress = (event: {keyCode: KeyCode}): void => {
        if (event.keyCode === KeyCode.ESC && this.state.open) this.setState(() => ({open: false}));
    };
}

export default ClickableDropdown;
