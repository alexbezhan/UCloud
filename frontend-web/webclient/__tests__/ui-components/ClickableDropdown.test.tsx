import * as React from "react";
import {create} from "react-test-renderer";
import {ThemeProvider} from "styled-components";
import {theme} from "../../app/ui-components";
import ClickableDropdown from "../../app/ui-components/ClickableDropdown"

describe("ClickableDropdown", () => {
    it("Closed clickable dropdown", () => {
        expect(create(
            <ThemeProvider theme={theme}>
                <ClickableDropdown
                    trigger={<button />}
                >
                    child
                </ClickableDropdown>
            </ThemeProvider>
        )).toMatchSnapshot();
    });
});
