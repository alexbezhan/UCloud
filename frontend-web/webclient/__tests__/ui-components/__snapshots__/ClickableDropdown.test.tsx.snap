// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickableDropdown Closed clickable dropdown 1`] = `
.c1 {
  cursor: pointer;
}

.c0 {
  position: relative;
  display: inline-block;
}

.c0:hover > div {
  display: block;
}

<div
  className="c0"
  data-tag="dropdown"
>
  <span
    className="c1"
    cursor="pointer"
    onClick={[Function]}
  >
    <button />
  </span>
</div>
`;
